popdat/src/distribute_people.rs

```diff
@@ -8,8 +11,57 @@ pub fn assign_people_to_houses(
     areas: Vec<CensusArea>,
     map: &Map,
     rng: &mut XorShiftRng,
-    config: &Config,
+    _config: &Config,
 ) -> Vec<CensusPerson> {
-    // TODO We should generalize the approach of distribute_residents from importer/src/berlin.rs
-    todo!()
+    let mut people = Vec::new();
+
+    for area in areas {
+        let bldgs: Vec<map_model::BuildingID> = map
+            .all_buildings()
+            .into_iter()
+            .filter(|b| area.polygon.contains_pt(b.label_center) && b.bldg_type.has_residents())
+            .map(|b| b.id)
+            .collect();
+
+        // If the area is partly out-of-bounds, then scale down the number of residents linearly
+        // based on area of the overlapping part of the polygon.
+        let pct_overlap = Polygon::union_all(area.polygon.intersection(map.get_boundary_polygon()))
+            .area()
+            / area.polygon.area();
+        let num_residents = (pct_overlap * (area.total_population as f64)) as usize;
+        debug!(
+            "Distributing {} residents to {} buildings. {}% of this area overlapped with the map, \
+             scaled residents accordingly.",
+            prettyprint_usize(num_residents),
+            prettyprint_usize(bldgs.len()),
+            (pct_overlap * 100.0) as usize
+        );
+
+        // How do you randomly distribute num_residents into some buildings?
+        // https://stackoverflow.com/questions/2640053/getting-n-random-numbers-whose-sum-is-m
+        // TODO Problems:
+        // - Because of how we round, the sum might not exactly be num_residents
+        // - This is not a uniform distribution, per stackoverflow
+        // - Larger buildings should get more people
```

Comment:
I can work on this - porting over the functionality from proletariat robot . Or have you already started?

Reply:
Go for it! I don't think prolet rob handles this? This is the original code from importing Berlin. I never dove into a proper algorithm for doing this, besides glancing at SO a bit.

Reply:
> Larger buildings should get more people

Specifically this part I was commenting on. You can see some of the logic used by proletariat robot here: https://github.com/dabreegster/abstreet/blob/master/map_model/src/make/buildings.rs#L234

It's currently in map_model/src/make/buildings.rs, but I'm not sure that makes sense.

Reply:
Ah, right. It's indeed awkward to have these parameters hardcoded at this layer. The main reason for putting them there is that they look at the OSM tags as input, and storing all of the building tags really increases the file size of the final maps. In lieu of the different emerging approaches for figuring out number of residents per building, I'm not sure how to restructure all of this cleanly. Should the number live in map_model at all?

Reply:
My thinking is that map_model should mostly reflect only "what's in OSM", and nothing more. With that interpretation, I'd think it makes sense to replace the inferred "number of people" associated data instead with the raw square footage values. The raw square footage is an uninterpreted fact from OSM, not inferred.

Then the activity model could use the raw square footage however it wants - in the census backed model, to distribute a known populations, and when population data isn't available, the model could guess at the population based on the square footage.

Does that sound reasonable?

Reply:
> My thinking is that map_model should mostly reflect only "what's in OSM", and nothing more
+1... although what about `match_parcels_to_buildings` in `importer/src/seattle.rs`? Where should that data live? Using a separate file is quite a hassle, because anywhere that loads the map and needs this data also has to load another file with the parcel counts. The one use of this right now is the experiment, but the 15 min tool will use it too soon (to count the number of people in the isochrone). Loading an extra file is a hassle on the web, because we have to chain `MapLoader` with another `FileLoader`...

> Then the activity model could use the raw square footage however it wants 

Should we store sq footage in `Building`, or just `levels` and keep using `polygon.area`?

How do we keep preserving the more specific building type tags like `apartment`? Should there be more cases of `BuildingType`, but just stop including this guess for number of people?